src/sentry/digests/utilities.py

```diff
@@ -23,3 +24,134 @@ def get_digest_metadata(digest):
+    """
+    get_personalized_digests(project: Project, digest: Digest, users: Set[User]) -> Iterator[User, Digest]
+    """
+    # TODO(LB): I Know this is inefficent.
+    # In the case that ProjectOwnership does exist, I do the same query twice.
+    # Once with this statement and again with the call to ProjectOwnership.get_actors()
+    # Will follow up with another PR to reduce the number of queries.
+    if ProjectOwnership.objects.filter(project_id=project_id).exists():
+        events = get_events_from_digest(digest)
+        events_by_actor = build_events_by_actor(project_id, events)
+        events_by_user = convert_actors_to_user_set(events_by_actor, user_ids)
+        for user_id in user_ids:
+            yield user_id, build_custom_digest(digest, events_by_user[user_id])
+    else:
+        for user_id in user_ids:
+            yield user_id, digest
+
+
+def sort_records(records):
+    """
+    Sorts records for fetch_state method
+    fetch_state is expecting these records to be ordered from newest to oldest
+    """
+    return sorted(records, key=lambda r: r.value.event.datetime, reverse=True)
+
+
+def get_events_from_digest(digest):
+    events = []
+    for rule_groups in six.itervalues(digest):
+        for group_records in six.itervalues(rule_groups):
+            events.append(group_records[0].value.event)
+    return set(events)
+
+
+def build_custom_digest(original_digest, events):
+    """
+    build_custom_digest(original_digest: Digest, user_id: Int, events_by_users: Map[User_Id:Set(Events)]) -> Digest
+    """
+    user_digest = OrderedDict()
+    for rule, rule_groups in six.iteritems(original_digest):
+        user_rule_groups = OrderedDict()
+        for group, group_records in six.iteritems(rule_groups):
+            user_group_records = [
+                record for record in group_records
+                if record.value.event in events
+            ]
+            if user_group_records:
+                user_rule_groups[group] = sort_records(user_group_records)
+        if user_rule_groups:
+            user_digest[rule] = user_rule_groups
+    return user_digest
+
+
+def build_events_by_actor(project_id, events):
+    """
+    build_events_by_actor(project_id: Int, events: Set(Events)) -> Map[Actor:Set(Events)]
+    """
+    events_by_actor = {}
+    for event in events:
+        # TODO(LB): I Know this is inefficent.
+        # ProjectOwnership.get_owners is O(n) queries and I'm doing that O(len(events)) times
+        # I will create a follow-up PR to address this method's efficency problem
+        # Just wanted to make as few changes as possible for now.
+        actors, __ = ProjectOwnership.get_owners(project_id, event.data)
+        if actors == ProjectOwnership.Everyone:
+            actors = [actors]
+        for actor in actors:
+            if actor in events_by_actor:
+                events_by_actor[actor].add(event)
+            else:
+                events_by_actor[actor] = set([event])
```

Comment:
I like the idea due to the fact that the code is more streamlined, but I assume it's really not materially different. You pass the check to the set implicitly, correct?.

Reply:
I'm not sure I understand this part of the question

> You pass the check to the set implicitly, correct?.

but otherwise yes, it's not materially different. It's beneficial in that it removes a class of possible errors (accidentally overwriting values like we encountered before.)